File: hello-websocket-ts/server/ws_server.ts
// Hello WebSocket Protocol - TypeScript Server
import { WebSocketServer, WebSocket } from 'ws';
import { randomUUID } from 'crypto';
import {
  PORT, SERVER_LANG,
  MSG_HELLO, MSG_BONJOUR, MSG_ECHO_REQUEST, MSG_ECHO_RESPONSE,
  MSG_KISS_REQUEST, MSG_KISS_RESPONSE, MSG_PING, MSG_PONG,
  MSG_TIME_NOTIFICATION, MSG_RANDOM_NUMBER, MSG_HASH_RESPONSE,
  MSG_DISCONNECT, MSG_ERROR,
  ERR_DECODE, ERR_UNKNOWN_MSG_TYPE,
  PING_INTERVAL_MS, SESSION_TIMEOUT_MS, TIME_INTERVAL_MS, KISS_INTERVAL_MS,
  encodeMessage, decodeMessage, log, nowMs, nowISO, hashNumber,
  type Message,
} from '../common/codec.js';

interface Session {
  userId: string;
  sessionId: string;
  clientLanguage: string;
  lastPongTs: bigint;
  timers: NodeJS.Timeout[];
}

const wss = new WebSocketServer({ port: PORT });

wss.on('connection', (ws: WebSocket, req) => {
  const userId = req.headers['userid'] as string || `ts-${randomUUID().substring(0, 8)}`;
  const session: Session = {
    userId,
    sessionId: randomUUID(),
    clientLanguage: 'unknown',
    lastPongTs: nowMs(),
    timers: [],
  };

  log('ws-server', `[${userId}] session+`);

  // Background: PING every 1s
  session.timers.push(setInterval(() => {
    if (ws.readyState === WebSocket.OPEN) {
      ws.send(encodeMessage({ type: MSG_PING, timestampMs: nowMs() }));
    }
  }, PING_INTERVAL_MS));

  // Background: TIME_NOTIFICATION every 5s
  session.timers.push(setInterval(() => {
    if (ws.readyState === WebSocket.OPEN) {
      ws.send(encodeMessage({ type: MSG_TIME_NOTIFICATION, timestampMs: nowMs(), iso8601: nowISO() }));
    }
  }, TIME_INTERVAL_MS));

  // Background: KISS_REQUEST every 5s
  session.timers.push(setInterval(() => {
    if (ws.readyState === WebSocket.OPEN) {
      ws.send(encodeMessage({
        type: MSG_KISS_REQUEST,
        osName: process.platform,
        osVersion: 'unknown',
        osRelease: 'unknown',
        osArch: process.arch,
      }));
    }
  }, KISS_INTERVAL_MS));

  // Background: timeout check every 5s
  session.timers.push(setInterval(() => {
    const last = session.lastPongTs;
    if (nowMs() - last > BigInt(SESSION_TIMEOUT_MS)) {
      log('ws-server', `[${userId}] session timeout`);
      ws.close();
    }
  }, 5000));

  ws.on('message', (data: Buffer) => {
    let msg: Message;
    try {
      msg = decodeMessage(data);
    } catch (e) {
      log('ws-server', `Decode error: ${(e as Error).message}`);
      ws.send(encodeMessage({ type: MSG_ERROR, errorCode: ERR_DECODE, errorMessage: (e as Error).message }));
      return;
    }

    switch (msg.type) {
      case MSG_HELLO:
        session.clientLanguage = msg.clientLanguage;
        log('ws-server', `HELLO from ${msg.clientLanguage}, session=${session.sessionId}, time=${nowMs()}`);
        ws.send(encodeMessage({ type: MSG_BONJOUR, serverLanguage: SERVER_LANG }));
        break;

      case MSG_ECHO_REQUEST:
        log('ws-server', `ECHO_REQUEST id=${msg.echoId} meta=${msg.echoMeta} data=${msg.echoData}`);
        ws.send(encodeMessage({
          type: MSG_ECHO_RESPONSE,
          echoStatus: 200,
          echoResults: [{
            idx: nowMs(),
            type: 0,
            kv: {
              id: msg.echoId.toString(),
              idx: msg.echoData,
              data: msg.echoData,
              meta: session.clientLanguage,
            },
          }],
        }));
        break;

      case MSG_KISS_RESPONSE:
        log('ws-server', `KISS_RESPONSE lang=${msg.kissLanguage} enc=${msg.kissEncoding} tz=${msg.kissTimeZone}`);
        break;

      case MSG_PONG:
        session.lastPongTs = msg.timestampMs;
        log('ws-server', `PONG ts=${msg.timestampMs}`);
        break;

      case MSG_RANDOM_NUMBER:
        log('ws-server', `RANDOM_NUMBER id=${msg.randomId} number=${msg.randomNumber}`);
        const hash = hashNumber(msg.randomNumber);
        ws.send(encodeMessage({ type: MSG_HASH_RESPONSE, randomId: msg.randomId, hashHex: hash }));
        log('ws-server', `HASH_RESPONSE id=${msg.randomId} hash=${hash}`);
        break;

      case MSG_DISCONNECT:
        log('ws-server', `DISCONNECT reason=${msg.disconnectReason}`);
        ws.close();
        break;

      case MSG_ERROR:
        log('ws-server', `ERROR code=${msg.errorCode} msg=${msg.errorMessage}`);
        break;

      default:
        log('ws-server', `Unknown message type: 0x${msg.type.toString(16)}`);
        ws.send(encodeMessage({ type: MSG_ERROR, errorCode: ERR_UNKNOWN_MSG_TYPE, errorMessage: `unknown type 0x${msg.type.toString(16)}` }));
    }
  });

  ws.on('close', () => {
    session.timers.forEach(t => clearInterval(t));
    log('ws-server', `[${userId}] session-`);
  });

  ws.on('error', (err: Error) => {
    log('ws-server', `Error: ${err.message}`);
  });
});

wss.on('listening', () => {
  log('ws-server', `Starting TypeScript WebSocket server on port ${PORT}`);
});

process.on('SIGINT', () => {
  log('ws-server', 'Shutting down...');
  wss.close();
  process.exit(0);
});

File: hello-websocket-cpp/server/ws_server.cpp
// Hello WebSocket Protocol - C++ Server
// Implements the full PROTOCOL.md server lifecycle: handshake, background tasks,
// echo, kiss, ping/pong, time broadcast, random/hash, disconnect.
#include "../common/codec.hpp"
#include "../common/ws.hpp"
#include <thread>
#include <mutex>
#include <atomic>
#include <sstream>
#include <cstring>
#include <cstdlib>

using namespace hws;

// Forward declaration
void handleClient(socket_t clientSock, const std::string& userId);

int main() {
    const char* portEnv = std::getenv("WS_PORT");
    uint16_t p = PORT;
    if (portEnv) {
        int v = atoi(portEnv);
        if (v > 0) p = static_cast<uint16_t>(v);
    }

    log("ws-server", "Starting C++ WebSocket server on port " + std::to_string(p));

    WSServer server;
    if (!server.listen(p)) {
        log("ws-server", "Failed to bind port " + std::to_string(p));
        return 1;
    }

    log("ws-server", "Listening on 0.0.0.0:" + std::to_string(p));

    while (true) {
        std::string userId;
        socket_t clientSock = server.accept(userId);
        if (clientSock == HWS_INVALID_SOCKET) continue;

        std::thread([clientSock, userId]() {
            handleClient(clientSock, userId);
        }).detach();
    }

    server.close();
    return 0;
}

void handleClient(socket_t clientSock, const std::string& userId) {
    std::string sessionId = std::to_string(nowMs());
    std::string clientLanguage = "unknown";

    std::atomic_flag sendLock = ATOMIC_FLAG_INIT;
    std::atomic<bool> active(true);
    std::atomic<int64_t> lastPongTs(nowMs());

    log("ws-server", "[" + userId + "] session+");

    // Helper to send binary frame (thread-safe)
    auto sendFrame = [&](const std::vector<uint8_t>& data) -> bool {
        while (sendLock.test_and_set(std::memory_order_acquire)) {}
        bool result = false;
        if (active) result = wsSendBinary(clientSock, data);
        sendLock.clear(std::memory_order_release);
        return result;
    };

    // Background: PING every 1s
    std::thread pingThread([&]() {
        while (active) {
            std::this_thread::sleep_for(std::chrono::milliseconds(PING_INTERVAL_MS));
            if (!active) break;
            Message m{};
            m.type = MSG_PING;
            m.timestampMs = nowMs();
            if (!sendFrame(m.encode())) { active = false; break; }
        }
    });

    // Background: TIME_NOTIFICATION every 5s
    std::thread timeThread([&]() {
        while (active) {
            std::this_thread::sleep_for(std::chrono::milliseconds(TIME_INTERVAL_MS));
            if (!active) break;
            Message m{};
            m.type = MSG_TIME_NOTIFICATION;
            m.timestampMs = nowMs();
            m.iso8601 = nowISO();
            if (!sendFrame(m.encode())) { active = false; break; }
        }
    });

    // Background: KISS_REQUEST every 5s
    std::thread kissThread([&]() {
        while (active) {
            std::this_thread::sleep_for(std::chrono::milliseconds(KISS_INTERVAL_MS));
            if (!active) break;
            Message m{};
            m.type = MSG_KISS_REQUEST;
#ifdef _WIN32
            m.osName = "Windows";
            m.osVersion = "unknown";
            m.osRelease = "unknown";
            m.osArch = "AMD64";
#else
            m.osName = "Linux";
            m.osVersion = "unknown";
            m.osRelease = "unknown";
            m.osArch = "x86_64";
#endif
            if (!sendFrame(m.encode())) { active = false; break; }
        }
    });

    // Background: timeout check every 5s
    std::thread timeoutThread([&]() {
        while (active) {
            std::this_thread::sleep_for(std::chrono::seconds(5));
            if (!active) break;
            if (nowMs() - lastPongTs.load() > static_cast<int64_t>(SESSION_TIMEOUT_MS)) {
                log("ws-server", "[" + userId + "] session timeout");
                active = false;
                break;
            }
        }
    });

    // Receive loop — wrapped so threads are ALWAYS joined before return
    try {
        while (active) {
            WSFrame frame = wsRecvFrame(clientSock);
            if (!frame.isValid || frame.isClose) break;

            Message msg;
            try {
                msg = decodeMessage(frame.payload.data(), frame.payload.size());
            } catch (const std::exception& e) {
                Message err{};
                err.type = MSG_ERROR;
                err.errorCode = ERR_DECODE;
                err.errorMessage = e.what();
                sendFrame(err.encode());
                continue;
            }

            switch (msg.type) {
                case MSG_HELLO:
                    clientLanguage = msg.clientLanguage;
                    log("ws-server", "HELLO from " + clientLanguage + ", session=" + sessionId);
                    {
                        Message bonjour{};
                        bonjour.type = MSG_BONJOUR;
                        bonjour.serverLanguage = SERVER_LANG;
                        sendFrame(bonjour.encode());
                    }
                    break;

                case MSG_ECHO_REQUEST:
                    log("ws-server", "ECHO_REQUEST id=" + std::to_string(msg.echoId));
                    {
                        Message resp{};
                        resp.type = MSG_ECHO_RESPONSE;
                        resp.echoStatus = 200;
                        EchoResult r;
                        r.idx = nowMs();
                        r.type = 0;
                        r.kv["id"] = std::to_string(msg.echoId);
                        r.kv["data"] = msg.echoData;
                        r.kv["meta"] = msg.echoMeta;
                        r.kv["lang"] = clientLanguage;
                        resp.echoResults.push_back(r);
                        sendFrame(resp.encode());
                    }
                    break;

                case MSG_KISS_RESPONSE:
                    log("ws-server", "KISS_RESPONSE lang=" + msg.kissLanguage + " enc=" + msg.kissEncoding + " tz=" + msg.kissTimeZone);
                    break;

                case MSG_PONG:
                    lastPongTs.store(msg.timestampMs);
                    break;

                case MSG_RANDOM_NUMBER:
                    log("ws-server", "RANDOM_NUMBER id=" + std::to_string(msg.randomId) + " number=" + std::to_string(msg.randomNumber));
                    {
                        std::string hash = hashNumber(msg.randomNumber);
                        Message resp{};
                        resp.type = MSG_HASH_RESPONSE;
                        resp.randomId = msg.randomId;
                        resp.hashHex = hash;
                        sendFrame(resp.encode());
                    }
                    break;

                case MSG_DISCONNECT:
                    log("ws-server", "DISCONNECT reason=" + msg.disconnectReason);
                    active = false;
                    break;

                case MSG_ERROR:
                    log("ws-server", "ERROR code=" + std::to_string(msg.errorCode) + " msg=" + msg.errorMessage);
                    break;

                default: {
                    std::ostringstream ss;
                    ss << std::hex << static_cast<int>(msg.type);
                    log("ws-server", "Unknown message type: 0x" + ss.str());
                    Message err{};
                    err.type = MSG_ERROR;
                    err.errorCode = ERR_UNKNOWN_MSG_TYPE;
                    err.errorMessage = "unknown type 0x" + ss.str();
                    sendFrame(err.encode());
                }
            }
        }
    } catch (const std::exception& e) {
        log("ws-server", "[" + userId + "] exception: " + e.what());
    } catch (...) {
        log("ws-server", "[" + userId + "] unknown exception");
    }

    // Cleanup: always join threads before returning
    active = false;
    hws_close(clientSock);  // unblock any blocked sends in background threads
    if (pingThread.joinable()) pingThread.join();
    if (timeThread.joinable()) timeThread.join();
    if (kissThread.joinable()) kissThread.join();
    if (timeoutThread.joinable()) timeoutThread.join();

    log("ws-server", "[" + userId + "] session-");
}
